File: Top_100_Liked_Questions/98._Validate_Binary_Search_Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    // iterative
    let stack = [[root, -Infinity, Infinity]];
    
    while (stack.length) {
        let [node, lb, rb] = stack.pop();
        
        if (!node) {
            continue;
        }
        
        if (lb >= node.val || node.val >= rb) {
            return false;
        }
        
        stack.push([node.left, lb, node.val]);
        stack.push([node.right, node.val, rb]);
    }
    
    return true;
}


var isValidBST2 = function(root) {
    // recursive
    let helper = function(node, lb, rb) {
        if (!node) {
            return true;
        }
        
        return lb < node.val && node.val < rb && helper(node.left, lb, node.val) && helper(node.right, node.val, rb);
    }
    
    return helper(root, -Infinity, Infinity);
};

File: Top_100_Liked_Questions/98._Validate_Binary_Search_Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        # iterative method
        stack = [(root, float('-inf'), float('inf'))]
        while stack:
            node, lb, rb = stack.pop()
            if not node:
                continue
                
            if not lb < node.val < rb:
                return False
            
            stack.append((node.left, lb, node.val))
            stack.append((node.right, node.val, rb))
        
        return True
    
    
    def isValidBST2(self, root: TreeNode) -> bool:
        # recursive method
        # needs to fall between ranges
        def helper(node, leftBound, rightBound):
            if not node:
                return True
            
            return leftBound < node.val < rightBound and helper(node.left, leftBound, node.val) and helper(node.right, node.val, rightBound)
        
        return helper(root, float('-inf'), float('inf'))